Add comprehensive benchmarking suite

dvmhmdsd opened the pull request:
## Motivation
As discussed here #526 for prformance optimiztions, that there should be a benchmarking infrastructure:

- Run before and after changes to measure impact
•  Execute on every Node.js version in the test matrix  
•  Prevent future performance regressions
•  Provide data-driven evidence for performance claims

dvmhmdsd committed:
feat(benchmark): add comprehensive benchmarking suite for qs library

- Implemented parsing benchmarks for various query string formats including simple, nested, and array queries.
- Added stringifying benchmarks for different object structures and array formats.
- Created a runner to manage benchmark execution and results reporting.
- Introduced fixtures for generating test data patterns for benchmarking.
- Added functionality to compare current results against baseline benchmarks.
- Implemented options for saving current results as baseline and generating summary reports.
- Enhanced command-line interface for running benchmarks with various options.

ljharb commented on benchmark/package.json:
```diff
@@ -0,0 +1,19 @@
+  "dependencies": {
+    "benchmark": "^2.1.4",
+    "table": "^6.8.1",
+    "chalk": "^4.1.2"
```
let's use `colors@=1.4.0` instead

ljharb commented on benchmark/README.md:
```diff
@@ -0,0 +1,206 @@
+## Features
+
+- 🚀 **Comprehensive Test Coverage**: Parse and stringify operations across various scenarios
+- 📊 **Cross-Node Version Testing**: Automated testing on Node.js 16, 18, 20, and 22
```
the test matrix goes down to node 0.8.

ljharb reviewed (changes requested):
the root package.json's publishConfig.ignore will need `"benchmark"` added in package.json as well.

Also, can we add a GHA workflow that runs these on every push/PR, and fails if there's a big enough slowdown?

dvmhmdsd committed:
fix(package): ignore benchmark directory in publishConfig

dvmhmdsd committed:
refactor: replace chalk with colors

dvmhmdsd committed:
fix(docs): update Node.js version range in benchmarking features

dvmhmdsd commented:
I added most of the comments, will find a way for the GHA @ljharb

dvmhmdsd committed:
feat(benchmark): add performance benchmarking workflow and comparison script

ljharb commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,81 @@
+        with:
+          fetch-depth: 0  # Get full history for comparison
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20'
+          cache: 'npm'
+
+      - name: Install dependencies
+        run: npm ci
```
please use ljharb/actions/node/install

ljharb commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,81 @@
+          git checkout origin/${{ github.base_ref }}
+          # Install dependencies (in case they changed)
+          npm ci
+          cd benchmark && npm ci && cd ..
```
please use `npm install` throughout; packages shouldn't have lockfiles and this one doesn't.

ljharb reviewed (commented):
this seems to be only running benchmarks in node 20.

dvmhmdsd commented:
> this seems to be only running benchmarks in node 20.

what versions should it run on ?

dvmhmdsd committed:
fix(workflow): update Node.js setup and dependency installation in benchmark workflow

ljharb commented:
> Execute on every Node.js version in the test matrix

all of them

dvmhmdsd commented:
Currently, the benchmarking scripts use some modern features (e.g., Array.from, async/await) and rely on tools that are not compatible with 0.8. So, is compatibility with Node 0.8 for the benchmarking infrastructure is a hard requirement ? In addition, some libraries i used for benchmarking is not compatible with node 0.8, for example, **Benchmark.js** which is use only supports node 0.10. @ljharb

ljharb commented:
I can definitely live with only going down to 0.10. https://npmjs.com/array.from should cover that one, and async/await can be done with Promises instead, which are polyfillable.

dvmhmdsd committed:
feat: translate es6 to es5 & enhance benchmarking workflows

ljharb commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,52 @@
+
+    strategy:
+      matrix:
+        node-version: [0.8, 0.9, 0.10, 0.11, 0.12, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
```
this is fine for iterating on the PR, but before merging this should change to use ljharb/actions/matrix with type "majors" so it automatically includes new node majors as they come out.

dvmhmdsd committed:
feat(benchmark): enhance performance benchmarking workflow and update query retrieval logic

ljharb commented on benchmark/polyfills.js:
```diff
@@ -0,0 +1,187 @@
+ * This file provides implementations for features not available in older Node versions
+ */
+
+// Polyfill for Array.from (ES6)
+if (!Array.from) {
```
please use https://npmjs.com/array.from and `require('array.from/auto')` for this.

ljharb commented on benchmark/polyfills.js:
```diff
@@ -0,0 +1,187 @@
+}
+
+// Polyfill for String.prototype.repeat (ES6)
+if (!String.prototype.repeat) {
+    String.prototype.repeat = function (count) {
```
https://npmjs.com/string.prototype.repeat

ljharb commented on benchmark/polyfills.js:
```diff
@@ -0,0 +1,187 @@
+
+// Polyfill for Object.fromEntries (ES2019)
+if (!Object.fromEntries) {
+    Object.fromEntries = function (entries) {
```
https://npmjs.com/object.fromentries

ljharb commented on benchmark/polyfills.js:
```diff
@@ -0,0 +1,187 @@
+
+// Polyfill for Object.entries (ES2017)
+if (!Object.entries) {
+    Object.entries = function (obj) {
```
https://npmjs.com/object.entries

ljharb commented on benchmark/polyfills.js:
```diff
@@ -0,0 +1,187 @@
+    Object.entries = function (obj) {
+        var ownProps = Object.keys(obj);
+        var i = ownProps.length;
+        var resArray = new Array(i);
```
as a side note, never use `new Array`, ever, as it makes a permanently slow object.

ljharb commented on benchmark/polyfills.js:
```diff
@@ -0,0 +1,187 @@
+// Simplified polyfill for Promise (ES6) for Node.js 0.8+
+// For complex usage, consider using a full Promise library like Q or Bluebird
+if (typeof Promise === 'undefined') {
+    global.Promise = function (executor) {
```
https://npmjs.com/es6-shim (which may cover string repeat already)

ljharb commented on benchmark/polyfills.js:
```diff
@@ -0,0 +1,187 @@
+    return originalMkdirSync.apply(this, arguments);
+};
+
+// Ensure Array includes method (ES7)
+if (!Array.prototype.includes) {
+    Array.prototype.includes = function (searchElement, fromIndex) {
```
https://npmjs.com/array-includes

ljharb commented on benchmark/polyfills.js:
```diff
@@ -0,0 +1,187 @@
+    };
+}
+
+module.exports = {
+    loaded: true
+};
```
i don't think this is needed at all?
```suggestion
```

ljharb commented on benchmark/runner.js:
```diff
@@ -0,0 +1,224 @@
+require('./polyfills');
+
+function table(data) {
+    return data.map(row => row.join('\t')).join('\n');
```
you'll need to avoid any post-ES5 syntax, including arrow functions

ljharb commented on benchmark/runner.js:
```diff
@@ -0,0 +1,224 @@
+    }
+};
+
+module.exports = BenchmarkRunner;
+
+module.exports = BenchmarkRunner;
+
```
```suggestion
module.exports = BenchmarkRunner;
```

ljharb commented on benchmark/runner.js:
```diff
@@ -0,0 +1,224 @@
+
+        console.log(table(comparisonData));
+    } catch (error) {
+        console.log(colors.red('❌ Error comparing with baseline: ' + error.message));
```
```suggestion
        console.error(colors.red('❌ Error comparing with baseline: ' + error.message));
```

ljharb commented on benchmark/stringify.js:
```diff
@@ -0,0 +1,101 @@
+
+var BenchmarkRunner = require('./runner');
+var fixtures = require('./fixtures');
+var qs = require('../lib/index');
```
```suggestion
var qs = require('../lib');
```

ljharb commented on package.json:
```diff
@@ -77,12 +77,19 @@
         "readme": "evalmd README.md",
         "postlint": "eclint check $(git ls-files | xargs find 2> /dev/null | grep -vE 'node_modules|\\.git' | grep -v dist/)",
         "lint": "eslint --ext=js,mjs .",
+        "benchmark": "cd benchmark && npm install && node index.js",
+        "benchmark:parse": "cd benchmark && npm install && node index.js --parse",
+        "benchmark:stringify": "cd benchmark && npm install && node index.js --stringify",
+        "benchmark:baseline": "cd benchmark && npm install && node index.js --save-baseline",
+        "benchmark:compare": "cd benchmark && npm install && node index.js --baseline benchmark/baseline.json",
```
```suggestion
        "benchmark": "cd benchmark && npm install && node .",
        "benchmark:parse": "cd benchmark && npm install && node . --parse",
        "benchmark:stringify": "cd benchmark && npm install && node . --stringify",
        "benchmark:baseline": "cd benchmark && npm install && node . --save-baseline",
        "benchmark:compare": "cd benchmark && npm install && node . --baseline benchmark/baseline.json",
```

also instead of all these `:foo` scripts, can these be `npm run benchmark -- --parse` etc?

ljharb commented on package.json:
```diff
@@ -77,12 +77,19 @@
+        "benchmark:stringify": "cd benchmark && npm install && node index.js --stringify",
+        "benchmark:baseline": "cd benchmark && npm install && node index.js --save-baseline",
+        "benchmark:compare": "cd benchmark && npm install && node index.js --baseline benchmark/baseline.json",
+        "benchmark:stress": "cd benchmark && npm install && STRESS_TEST=1 node index.js",
```
```suggestion
        "benchmark:stress": "STRESS_TEST=1 npm run benchmark",
```

dvmhmdsd committed:
Refactor benchmark suite and update results

- Deleted outdated benchmark result files to clean up the repository.
- Updated the `generateQuery` function in fixtures to handle default size parameter more gracefully.
- Added new dependencies for polyfills to support ES6+ features in older Node.js versions.
- Removed custom polyfills for Array.from, String.prototype.repeat, Object.fromEntries, and Object.entries, replacing them with well-maintained npm packages.
- Updated benchmark runner to improve code readability and maintainability.
- Adjusted benchmark commands in package.json for consistency and clarity.
- Added new benchmark result file with updated performance metrics for various parsing scenarios.

dvmhmdsd committed:
fix(benchmark): simplify benchmark script commands for improved readability

dvmhmdsd committed:
Fix formatting

ljharb commented on README.md:
````diff
@@ -314,14 +319,14 @@ To disable array parsing entirely, set `parseArrays` to `false`.
 ```javascript
 var noParsingArrays = qs.parse('a[]=b', { parseArrays: false });
-assert.deepEqual(noParsingArrays, { a: { '0': 'b' } });
+assert.deepEqual(noParsingArrays, { a: { 0: 'b' } });
````
did you perhaps accidentally autoformat this file? if so please revert it :-)

dvmhmdsd replied:
oh yea by accident

dvmhmdsd committed:
Revert "Fix formatting"

This reverts commit eb599f5801382eb7e971e2678e18ecf699e29e3b.

dvmhmdsd committed:
refactor(benchmark): update module exports to use explicit property assignment and change arrow functions to regular functions

dvmhmdsd committed:
fix(benchmark): change const to var for queries object in generateQuery function

dvmhmdsd commented:
Hi @ljharb 
Just following up on this PR — I’ve addressed all the comments and happy to make any further changes if needed. Please let me know if there's anything else blocking it from being merged. Thanks.

ljharb commented:
Sorry for the long delay. If you could rebase this and resolve the conflicts, I can give it a new review.
